pin/source.go

```diff
@@ -0,0 +1,34 @@
+package pin
+
+import (
+	cid "gx/ipfs/QmNp85zy9RLrQ5oQD4hPyS39ezrrXpcaa7R4Y9kxdWQLLQ/go-cid"
+)
+
+// Source is structure describing source of the pin
+type Source struct {
+	// Get if a function that will be called to get the pins for GCing
+	Get func() ([]*cid.Cid, error)
+	// Strict makes the GC fail if some objects can't be fetched during
+	// recursive traversal of the graph
+	Strict bool
+	// Direct marks the pinned object as the final object in the traversal
+	Direct bool
+	// Internal marks the pin source which recursive enumeration should be
+	// terminated by a direct pin
+	Internal bool
```

Comment:
Can we have better names for these? I'm still confused, even after reading the comments.

* I understand `Strict` and can't think of a better name.
* `Direct`, IIRC, means non-recursive. That comment is technically correct but leads one to believe that it's a terminal pin in a DAG (`Internal`). Maybe invert this and use `Recursive`?
* `Internal` -> `Terminal`?  Or, maybe, `Never` (i.e., never pin)? Technically, it's a pin that's "internal" to a recursive pin but the name doesn't indicate anything about it being a "terminal" pin.

Reply:
Also, "internal" is a rather big hammer. I guess it's a decent starting point (for, e.g., excluding large files globally) but, really, I'd like to be able to say, e.g., "pin A recursively but stop at X and Y" and have this pin *not* mess with other pins (that may want X/Y). Are pins like this expressible with this GC without major surgery? This isn't a reason to hold this PR (future work) up but we're going to want to be able to express constraints like this (and, eventually, arbitrary IPLD selectors for, e.g., ipfs-cluster).

Reply:
> Maybe invert this and use Recursive?

`Direct` works much nicer with default values (false).

> "pin A recursively but stop at X and Y"

This might be a bit hard and we will have to think carefully about it. If we are able to sort the selectors according to their reach (some may be super-sets of others, so may have parts in common but then select different objects).

Big optimization that comes from color set (it was already implemented before but not explicit) is that we don't need to enumerate parts of the tree that were already enumerated. 

> we're going to want to be able to express constraints like this

The color system allows for a lot of flexibility and having it here explicitly is IMO great step forward. Right now internal means "enumerate it last", which means it stops as elements that are already marked black (as they are supposedly fully enumerated but it is not the case in case of `Direct` pins).

Reply:
> `Direct` works much nicer with default values (false).
> Right now internal means "enumerate it last", which means it stops as elements that are already marked black (as they are supposedly fully enumerated but it is not the case in case of Direct pins).

Oh, now I understand. Direct is an overloaded term/variable. Direct pins act as both non-recursive pins *and* stop-pins for internal pins. Is that right? At the very least, this algorithm needs to be documented somewhere. Personally, I'd prefer something like:

```go
type PinKind int

const (
    PinRecursive PinKind = iota
    PinTerminal // Terminates an internal pin.
    PinInternal // Better name...
    PinSingle // one-of pins, add these last.
) // sort order is const order.

struct PinSource {
    Kind PinKind // Can only be one type.
    Strict bool
    Get GetFunc
}
```

> Big optimization that comes from color set (it was already implemented before but not explicit) is that we don't need to enumerate parts of the tree that were already enumerated

So, I know of a way to make this work but it will be a lot of work and will require storing extra data; we can deal with it later. The TL;DR is to maintain a (semi-accurate) topological sort of the DAG and operate over that. We'll have to see how much extra work maintaining this datastructure will be but it *will* make executing IPLD queries much faster (IMO, it makes sense to optimize for this).

However, that's not something we'll need to deal with now.

> Shift as in: Selector on node A is X and on node B is Y, they have both selected node C (the same node) and they both simplify to Z.

With topological sorting, I believe we can do this. Without it, we have the problem where we may have already marked C and it's children with respect to query X rooted at A by the time we get around to processing query Y rooted at B.